File: app/src/main/java/com/example/networkdiscovery02/NsdServerManager.java
package com.example.networkdiscovery02;

import android.content.Context;
import android.util.Log;

public class NsdServerManager {
    private static final String TAG = "NsdServerManagerWhy";
    private volatile static NsdServerManager nsdServerManager;
    private static NSDServer nsdServer;
    private Context context;

    public NsdServerManager(Context context){
        this.context=context;
    }

    public static NsdServerManager getInstance(Context context){
        if(nsdServerManager==null){
            synchronized (NsdServerManager.class){
                if (nsdServerManager==null) {
                    nsdServer=new NSDServer();
                    nsdServerManager = new NsdServerManager(context);
                }
            }
        }
        return nsdServerManager;
    }

    /**
     * Register for Nsd service
     * @param nsdServerName
     */
    public void registerNsdServer(String nsdServerName) {
        new Thread(new NsdServerRunnable(nsdServerName,nsdServer,context)).start();
    }

    /**
     * Unregister NsdServer
     */
    public void unRegisterNsdServer(){
        Log.e(TAG, "unRegisterNsdServer: " );
        nsdServer.stopNSDServer();
    }


}

File: app/src/main/java/com/example/networkdiscovery02/NsdServerManager.kt
package com.example.networkdiscovery02

import android.content.Context
import android.util.Log

class NsdServerManager(private val context: Context) {
    /**
     * Register for Nsd service
     * @param nsdServerName
     */
    fun registerNsdServer(nsdServerName: String?) {
        Thread(NsdServerRunnable(nsdServerName!!, nsdServer!!, context)).start()
    }

    /**
     * Unregister NsdServer
     */
    fun unRegisterNsdServer() {
        Log.e(TAG, "unRegisterNsdServer: ")
        nsdServer!!.stopNSDServer()
    }

    companion object {
        private const val TAG = "NsdServerManagerWhy"

        @Volatile
        private var nsdServerManager: NsdServerManager? = null
        private var nsdServer: NSDServer? = null
        fun getInstance(context: Context): NsdServerManager? {
            if (nsdServerManager == null) {
                synchronized(NsdServerManager::class.java) {
                    if (nsdServerManager == null) {
                        nsdServer = NSDServer()
                        nsdServerManager = NsdServerManager(context)
                    }
                }
            }
            return nsdServerManager
        }
    }
}